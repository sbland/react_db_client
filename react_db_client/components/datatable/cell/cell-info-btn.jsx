import React from 'react';
import PropTypes from 'prop-types';

export const CellInfoBtn = ({ styles, message }) => {
  const handleTagButtonClick = () => {
    // TODO: Replace alert with notification
    if (message) alert(message.text);
  };

  const lhsTagStyle = {
    ...styles,
    width: '8px',
    fontSize: '1rem',
    lineHeight: '1rem',
    color: 'white',
    borderRadius: '0.3rem',
    fontWeight: 800,
    textAlign: 'center',
    cursor: 'pointer',
  };

  const lhsTagClass = [
    'button-reset',
    message ? 'popIn invalidRow' : 'hidden', // TODO: Move style to datatable css
  ].join(' ');

  return (
    <button
      type="button"
      className={lhsTagClass}
      style={lhsTagStyle}
      onClick={handleTagButtonClick}
    >
      {message ? '!' : ' '}
    </button>
  );
};

CellInfoBtn.propTypes = {
  styles: PropTypes.objectOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])).isRequired,
  message: PropTypes.string,
};

CellInfoBtn.defaultProps = {
  message: '',
}
