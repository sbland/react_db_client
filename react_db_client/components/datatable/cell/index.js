export { Cell } from './cell';
export { DataTableCellHoverWrap } from './cell-wrappers';
