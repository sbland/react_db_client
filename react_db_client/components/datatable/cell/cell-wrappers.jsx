import React from 'react';
import PropTypes from 'prop-types';
import { Emoji } from '@react_db_client/components.emoji';
// import '../_dataTable.scss';
// import { RightClickWrapper } from '@samnbuk/react_db_client.components.popup-menu';
import { DataTableCellReadOnly } from '@samnbuk/react_db_client.components.datatable.cell-types';

/**
 * Data Table cell wrap that handles cell hovering
 *
 * @param {*} { className, style, handleHover, children }
 */
export const DataTableCellHoverWrap = ({
  className,
  style,
  handleHover,
  // columnWidth,
  disabled,
}) => (
  <div
    className={`dataTableCell_wrap ${className} ${disabled ? 'disabled' : 'enabled'}`}
    onFocus={() => handleHover && handleHover(true)}
    onMouseEnter={() => handleHover && handleHover(true)}
    onMouseLeave={() => handleHover && handleHover(false)}
    onBlur={() => handleHover && handleHover(false)}
    style={{
      // width: '100%',
      // height: '100%',
      ...style,
    }}
  />
);

DataTableCellHoverWrap.propTypes = {
  className: PropTypes.string,
  style: PropTypes.object,
  handleHover: PropTypes.func,
  columnWidth: PropTypes.number,
  disabled: PropTypes.bool,
};

DataTableCellHoverWrap.defaultProps = {
  className: '',
  style: {},
  handleHover: () => {},
  columnWidth: 100,
  disabled: false,
};

/**
 *
 * @param {*} param0
 * @returns
 */
export const EditColumnCell = ({
  allowSelection,
  allowRowDelete,
  allowRowEditPanel,
  isSelected,
  rowIndex,
  handleRemoveFromSelection,
  handleAddToSelection,
  handleDeleteRow,
  handleEditPanelBtnClick,
  rowUid,
}) => {
  return (
    <>
      {allowSelection && (
        <input
          className="rowSelectionBox"
          type="checkbox"
          id={rowIndex}
          checked={isSelected || false}
          onChange={() =>
            isSelected
              ? handleRemoveFromSelection(rowUid, rowIndex)
              : handleAddToSelection(rowUid, rowIndex)
          }
        />
      )}
      {allowRowDelete && (
        <button
          type="button"
          className="rowDeleteBtn"
          onClick={() => {
            handleDeleteRow(rowUid, rowIndex);
          }}
        >
          <Emoji emoj="🗑️" label="Delete" />
        </button>
      )}
      {allowRowEditPanel && (
        <button
          type="button"
          onClick={() => {
            handleEditPanelBtnClick(rowUid, rowIndex);
          }}
        >
          <Emoji emoj="📖" label="Open" />
        </button>
      )}
    </>
  );
};

EditColumnCell.propTypes = {
  allowSelection: PropTypes.bool.isRequired,
  allowRowDelete: PropTypes.bool.isRequired,
  allowRowEditPanel: PropTypes.bool.isRequired,
  isSelected: PropTypes.bool.isRequired,
  rowIndex: PropTypes.number.isRequired,
  handleRemoveFromSelection: PropTypes.func.isRequired,
  handleAddToSelection: PropTypes.func.isRequired,
  handleDeleteRow: PropTypes.func.isRequired,
  handleEditPanelBtnClick: PropTypes.func.isRequired,
  rowUid: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
};

// export const CellRightClickWrapper = ({ readOnly, clearCell, setAsDefault, children }) => {
//   if (readOnly) {
//     return children || '';
//   }

//   return (
//     <RightClickWrapper
//       items={[
//         { uid: 'clearCell', label: 'Clear', onClick: clearCell },
//         {
//           uid: 'setDefault',
//           label: 'Set as Default',
//           onClick: setAsDefault,
//         },
//       ]}
//       //TODO: How do we make this generic?
//       popupRoot="root"
//     >
//       {children}
//     </RightClickWrapper>
//   );
// };

// CellRightClickWrapper.propTypes = {
//   readOnly: PropTypes.bool,
//   clearCell: PropTypes.func.isRequired,
//   setAsDefault: PropTypes.func.isRequired,
//   children: PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf(PropTypes.nodes)]).isRequired,
// };

// CellRightClickWrapper.defaultProps = {
//   readOnly: false,
// };

const defaultComponent = () => (props) => <DataTableCellReadOnly {...props} />;

/**
 * Data Table Cell
 * Selects the correct cell type based on headings
 * Adds a right click wrapper
 *
 * @param {*} {
 *   rowId,
 *   cellData,
 *   columnData,
 *   updateData,
 * }
 * @returns
 */
// export const DataTableDataCell = (props) => {
//   const {
//     columnData: { type, readOnly, defaultValue },
//     updateData,
//     rowId,
//     columnId,
//     isDisabled,
//     componentMap,
//   } = props;
//   // TODO: Make sure read only component can parse any data
//   // const cellComponent = useMemo(
//   //   () =>
//   //     readOnly ? (
//   //       <DataTableCellReadOnly {...props} />
//   //     ) : (
//   //       switchF(type, componentMap, defaultComponent)
//   //     ),
//   //   [type, componentMap]
//   // );

//   const CellComponent = useMemo(
//     () => switchF(type, componentMap, defaultComponent),
//     [type, componentMap, defaultComponent]
//   );

//   // TODO: Can this not also handle read only?
//   if (isDisabled) {
//     return (
//       <div style={{ height: '100%' }} className="disabled">
//         <CellComponent {...props} />
//       </div>
//     );
//   }

//   return (
//     <div style={{ height: '100%' }}>
//         <CellComponent {...props} />
//     </div>
//   );
// };

// DataTableDataCell.propTypes = {
//   rowId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
//   columnId: PropTypes.string.isRequired,
//   rowData: PropTypes.shape({}).isRequired,
//   columnData: PropTypes.shape({
//     uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
//     label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
//     type: PropTypes.string.isRequired,
//     action: PropTypes.func,
//     to: PropTypes.string,
//     readOnly: PropTypes.bool,
//     defaultValue: PropTypes.any,
//   }).isRequired,
//   updateData: PropTypes.func.isRequired,
//   componentMap: PropTypes.objectOf(PropTypes.elementType).isRequired,
//   isDisabled: PropTypes.bool,
// };

// DataTableDataCell.defaultProps = {
//   isDisabled: false,
// }