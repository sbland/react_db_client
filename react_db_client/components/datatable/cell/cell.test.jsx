import '@samnbuk/react_db_client.helpers.enzyme-setup';
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { sleep } from '@samnbuk/react_db_client.helpers.testing';
import { Cell } from './cell';
import { defaultComponentMap } from '@samnbuk/react_db_client.components.datatable.components';
import { DataTableDataCell } from './cell-wrappers';
import { DataTableCellText } from '@samnbuk/react_db_client.components.datatable.cell-types';
import {
  demoHeadingsData,
  demoTableData,
} from '@samnbuk/react_db_client.components.datatable.extras';
import {
  DataTableContext,
  dataTableDefaultConfig,
} from '@samnbuk/react_db_client.components.datatable.logic';

import {
  CustomSelectDropdown,
  DropdownItem,
} from '@samnbuk/react_db_client.components.custom-select-dropdown';
import { act } from 'react-dom/test-utils';

const DEMO_TABLE_DATA = Object.values(demoTableData);

const handleValueAccept = jest.fn();
const handleValueChange = jest.fn();
const handleValueReset = jest.fn();
// const handleDeleteRow = jest.fn();
const handleMoveFocusToTargetCell = jest.fn();
// const handleEditPanelBtnClick = jest.fn();
// const handleAddToSelection = jest.fn();
// const handleRemoveFromSelection = jest.fn();
const setNavigationMode = jest.fn();

const ROW_INDEX = 1;
const COLUMN_INDEX = 1;
const defaultProps = {
  columnIndex: COLUMN_INDEX,
  rowIndex: ROW_INDEX,
  style: {},
  className: '',
  tableData: DEMO_TABLE_DATA,
  headingsData: demoHeadingsData,
  methods: {
    handleValueAccept,
    handleValueChange,
    handleValueReset,
    handleMoveFocusToTargetCell,
    setNavigationMode,
  },
  currentFocusedRow: 0,
  currentFocusedColumn: 0,
  navigationMode: true,
  componentMap: defaultComponentMap(),
};

describe('DatatableCell', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  describe('Renders', () => {
    test('Ok', () => {
      mount(<Cell {...defaultProps} />);
    });
    test('Matches Snapshot', () => {
      const component = mount(
        <DataTableContext.Provider value={dataTableDefaultConfig}>
          <Cell {...defaultProps} />
        </DataTableContext.Provider>
      );
      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
    });
  });
  describe('Unit tests', () => {
    describe('Navigation', () => {
      /* Helper Utils */
      const getComponent = ({
        isFocused = true,
        navigationMode = true,
        rowIndex = ROW_INDEX,
        columnIndex = COLUMN_INDEX,
        override = {},
      } = {}) => {
        document.body.innerHTML = `
            <div id="wrap"></div>
    `;
        const wrap = document.getElementById('wrap');
        const props = isFocused
          ? {
              ...defaultProps,
              rowIndex,
              columnIndex,
              ...defaultProps.data,
              currentFocusedColumn: columnIndex,
              currentFocusedRow: rowIndex,
              navigationMode,
              ...override,
            }
          : { ...defaultProps, rowIndex, columnIndex };
        return mount(
          <DataTableContext.Provider value={dataTableDefaultConfig}>
            <Cell {...props} />
          </DataTableContext.Provider>,
          { attachTo: wrap }
        );
      };

      const pressKeyOnNavBtn = (c, key) => {
        const navigationBtn = c.find('.navigationButton');
        navigationBtn.simulate('keyDown', { key });
        c.update();
      };

      const enterEditMode = (c) => {
        const navigationBtn = c.find('.navigationButton');
        navigationBtn.simulate('keyDown', { key: 'Enter' });
        c.setProps({ ...c.props(), navigationMode: false });
        c.update();
      };

      const resetCell = (c) => {
        const cell = c.find(DataTableDataCell);
        cell.props().resetValue();
        c.update();
      };

      const acceptCell = (c) => {
        const cell = c.find(DataTableDataCell);
        act(() => {
          cell.props().acceptValue('New Val');
        });
      };

      // const checkIsInNavigationMode = (c) => {
      //   const dataTableCellData = c.find(DataTableDataCell);
      //   return dataTableCellData.props().focused;
      // };

      const checkIsInEditMode = (c) => {
        const dataTableCellData = c.find(DataTableDataCell);
        return dataTableCellData.props().editMode;
      };

      /* Test */
      let component;
      beforeEach(() => {
        component = getComponent();
      });

      afterEach(() => {
        component && component.unmount();
      });

      test('should focus on the navigation button if this cell is focused', async () => {
        component = getComponent({ isFocused: true });
        const navigationBtn = component.find('.navigationButton');
        navigationBtn.getDOMNode();
        await sleep(1000);

        const focusedElement = document.activeElement;
        expect(navigationBtn.getDOMNode()).toBe(focusedElement);
      });

      test('should not focus on the navigation button if this cell is not focused', () => {
        component = getComponent({ isFocused: false });
        const navigationBtn = component.find('.navigationButton');
        const focusedElement = document.activeElement;
        expect(navigationBtn.getDOMNode()).not.toBe(focusedElement);
      });

      test('should turn off navigation mode when pressing enter on focused cell', () => {
        component = getComponent({ isFocused: true });
        pressKeyOnNavBtn(component, 'Enter');
        expect(setNavigationMode).toHaveBeenCalledWith(false);
      });

      test('should call cell hover on next cell if we use arrow keys', () => {
        pressKeyOnNavBtn(component, 'ArrowRight');
        expect(handleMoveFocusToTargetCell).toHaveBeenCalledWith(ROW_INDEX, COLUMN_INDEX + 1);
        handleMoveFocusToTargetCell.mockClear();
        pressKeyOnNavBtn(component, 'ArrowUp');
        expect(handleMoveFocusToTargetCell).toHaveBeenCalledWith(ROW_INDEX - 1, COLUMN_INDEX);
      });

      test('should turn on edit mode when pressing enter on focused cell', () => {
        component = getComponent({ isFocused: true, columnIndex: 2 });
        expect(checkIsInEditMode(component)).toEqual(false);
        enterEditMode(component);
        expect(checkIsInEditMode(component)).toEqual(true);
      });

      test('should go back to navigation mode if reset value', () => {
        component = getComponent({ isFocused: true, navigationMode: false });
        resetCell(component);
        expect(setNavigationMode).toHaveBeenCalledWith(true);
        expect(checkIsInEditMode(component)).toEqual(false);
      });

      describe.skip('navigate type specific edge cases', () => {
        describe('button cell type', () => {
          test('should call cell action when pressing enter on button cell', () => {
            const columnData = {
              uid: 'uid',
              label: 'UID',
              type: 'button',
              columnId: 1,
              action: jest.fn(),
            };

            component = getComponent({
              isFocused: true,
              navigationMode: true,
              rowIndex: 0,
              columnIndex: 1,
              override: {},
              headingsData: [columnData],
            });

            const initSnap = toJson(component);
            expect(checkIsInEditMode(component)).toEqual(false);
            enterEditMode(component);
            component.update();
            console.log(toJson(component));
            expect(initSnap).toMatchDiffSnapshot(toJson(component));
            // expect(checkIsInEditMode(component)).toEqual(false);
            expect(columnData.action).toHaveBeenCalledWith(
              DEMO_TABLE_DATA[0].uid,
              DEMO_TABLE_DATA[0][columnData.uid],
              DEMO_TABLE_DATA[0]
            );
          });
        });

        describe('bool cell type', () => {
          test('should toggle bool cell when pressing enter on bool cell', () => {
            const columnData = {
              uid: 'toggle',
              label: 'Toggle',
              type: 'toggle',
            };

            component = getComponent({
              isFocused: true,
              navigationMode: true,
              rowIndex: 0,
              columnIndex: 1,
              override: {},
              dataOverride: { headingsData: [columnData] },
            });
            enterEditMode(component);
            component.update();
            expect(checkIsInEditMode(component)).toEqual(false);
            expect(handleValueAccept).toHaveBeenCalledWith(
              !DEMO_TABLE_DATA[0][columnData.uid],
              DEMO_TABLE_DATA[0].uid,
              0,
              columnData.uid
            );
          });
        });

        describe('text-long cell type', () => {
          test('should focus on inner text area if enter pressed on nav button', () => {
            component = getComponent({
              isFocused: true,
              navigationMode: true,
              rowIndex: ROW_INDEX,
              columnIndex: 2,
            });
            pressKeyOnNavBtn(component, 'Enter');
            const textCell = component.find(DataTableCellText).find('.cellInput-text');
            const focusedElement = document.activeElement;
            expect(textCell.getDOMNode()).toBe(focusedElement);
          });
        });

        describe('select cell type', () => {
          const columnData = {
            uid: 'selectionField',
            label: 'UID',
            type: 'select',
            options: [{ uid: 'a', label: 'A' }],
          };
          const rowIndex = 0;

          const getSelComponent = () =>
            getComponent({
              isFocused: true,
              navigationMode: true,
              rowIndex,
              columnIndex: 1,
              override: {},
              headingsData: [columnData],
            });
          test('should open selection options automatically when pressing enter on focused cell', () => {
            component = getSelComponent();
            enterEditMode(component);
            component.update();
            const customDropDown = component.find(CustomSelectDropdown);
            expect(customDropDown.props().isOpen).toEqual(true);
          });

          test('should close selection options when a option is selected', () => {
            component = getSelComponent();
            const newVal = 'zzz';
            enterEditMode(component);
            let customDropDown = component.find(CustomSelectDropdown);
            customDropDown.props().handleSelect(newVal);
            expect(setNavigationMode).toHaveBeenCalledWith(true);
            component.update();
            expect(checkIsInEditMode(component)).toEqual(false);
            customDropDown = component.find(CustomSelectDropdown);
            expect(customDropDown.props().isOpen).toEqual(false);
            expect(handleValueAccept).toHaveBeenCalledWith(newVal, 'a', rowIndex, columnData.uid);
          });

          test('should accept value when accepting an option', () => {
            component = getSelComponent();
            enterEditMode(component);
            const newVal = 'zzz';
            const customDropDown = component.find(CustomSelectDropdown);
            const firstOption = customDropDown.find(DropdownItem).first();
            firstOption.props().handleSelect(newVal);
            expect(handleValueAccept).toHaveBeenCalledWith(newVal, 'a', rowIndex, columnData.uid);
          });

          test('should accept value when clicking an option', () => {
            component = getSelComponent();
            enterEditMode(component);
            const customDropDown = component.find(CustomSelectDropdown);
            const firstOption = customDropDown.find(DropdownItem).first().find('button');
            firstOption.simulate('click');
            component.update();
            expect(handleValueAccept).toHaveBeenCalledWith('a', 'a', rowIndex, columnData.uid);
          });
        });
      });

      test('should show and focus on navigation button and hide cell if reset value deep', () => {
        component = getComponent({
          isFocused: true,
          navigationMode: true,
          rowIndex: ROW_INDEX,
          columnIndex: 2,
        });
        enterEditMode(component);
        // pressKeyOnNavBtn(component, 'Enter');
        const textCell = component.find(DataTableCellText).find('.cellInput-textarea');
        textCell.simulate('keyDown', { key: 'Escape' });
        expect(setNavigationMode).toHaveBeenCalledWith(true);
        const focusedElement = document.activeElement;
        const navigationBtn = component.find('.navigationButton');
        expect(navigationBtn.getDOMNode()).toBe(focusedElement);
      });

      test('should show navigation button and hide cell if accept value', () => {
        component = getComponent({ isFocused: true });
        enterEditMode(component);
        acceptCell(component);
        expect(setNavigationMode).toHaveBeenCalledWith(false);
      });

      test('should go to edit mode and focus on the input of the cell if we click on it', () => {
        component = getComponent({ isFocused: true });
        const navigationBtn = component.find('.navigationButton');
        navigationBtn.simulate('click');
        expect(setNavigationMode).toHaveBeenCalledWith(false);
      });
    });
  });
});

// NOTE: Move this elsewhere as we no longer use btn column
// describe('Btns cell', () => {
//   test('Matches Snapshot', () => {
//     const props = {
//       ...defaultProps,
//       data: {
//         ...defaultProps.data,
//       },
//     };
//     const config = {
//       ...dataTableDefaultConfig,
//       allowEditRow: true,
//       allowRowDelete: true,
//       hasBtnsColumn: true,
//       allowSelection: true,
//     };
//     const component = mount(
//       <DataTableContext.Provider value={config}>
//         <Cell {...props} />
//       </DataTableContext.Provider>
//     );
//     const tree = component.debug();
//     expect(tree).toMatchSnapshot();
//   });
// });
