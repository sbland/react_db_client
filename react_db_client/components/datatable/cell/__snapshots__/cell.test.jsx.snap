// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatatableCell Renders Matches Snapshot 1`] = `
<Cell
  className=""
  columnIndex={1}
  componentMap={
    Object {
      "bool": [Function],
      "button": [Function],
      "number": [Function],
      "select": [Function],
      "text": [Function],
      "textLong": [Function],
      "toggle": [Function],
      "undefined": [Function],
    }
  }
  currentFocusedColumn={0}
  currentFocusedRow={0}
  disabled={false}
  headingsData={
    Array [
      Object {
        "action": [Function],
        "label": "UID",
        "required": true,
        "styleRule": "$count<5",
        "type": "button",
        "uid": "uid",
        "unique": true,
      },
      Object {
        "label": "Name",
        "required": true,
        "type": "text",
        "uid": "name",
        "unique": true,
      },
      Object {
        "label": "description",
        "type": "textLong",
        "uid": "description",
      },
      Object {
        "def": 8,
        "label": "Count",
        "max": 10,
        "min": 2,
        "showTotals": true,
        "step": 1,
        "type": "number",
        "uid": "count",
      },
      Object {
        "defaultValue": 7,
        "label": "Def",
        "max": 10,
        "min": 1,
        "showTotals": true,
        "type": "number",
        "uid": "def",
      },
      Object {
        "def": "$def",
        "label": "Multiplier",
        "max": 10,
        "min": 1,
        "type": "number",
        "uid": "multiplier",
      },
      Object {
        "label": "Read Only",
        "readOnly": true,
        "type": "text",
        "uid": "readOnly",
      },
      Object {
        "hidden": true,
        "label": "Hidden",
        "type": "text",
        "uid": "hiddenDemo",
      },
      Object {
        "hidden": true,
        "label": "Hidden Number",
        "type": "number",
        "uid": "hiddenDemoNumber",
      },
      Object {
        "label": "Toggle",
        "type": "toggle",
        "uid": "toggle",
      },
      Object {
        "label": "ToggleR",
        "readOnly": true,
        "type": "toggle",
        "uid": "toggleR",
      },
      Object {
        "evaluateType": "number",
        "expression": "$count * $multiplier",
        "expressionReversed": "$count=$_ / $multiplier",
        "label": "Expression",
        "reverseExpression": "$count * $multiplier",
        "showTotals": true,
        "type": "number",
        "uid": "expression",
      },
      Object {
        "evaluateType": "string",
        "field": "select",
        "invert": true,
        "label": "Is Validated",
        "operator": "=",
        "target": "a",
        "type": "string",
        "uid": "validated",
      },
      Object {
        "evaluateType": "number",
        "expression": "$nothing",
        "label": "Invalid",
        "type": "number",
        "uid": "invalid",
      },
      Object {
        "evaluateType": "number",
        "expression": "$invalid * 2",
        "label": "InvalidB",
        "type": "number",
        "uid": "invalidb",
      },
      Object {
        "label": "Select",
        "options": Array [
          Object {
            "label": "A",
            "uid": "a",
          },
          Object {
            "label": "B",
            "uid": "b",
          },
          Object {
            "label": "C",
            "uid": "c",
          },
        ],
        "type": "select",
        "uid": "select",
      },
      Object {
        "collection": "rooms",
        "label": "Internal Room",
        "type": "entity",
        "uid": "internalRoom",
      },
      Object {
        "collection": "rooms",
        "label": "Long Heading which is really really long",
        "type": "entity",
        "uid": "longHeading",
      },
      Object {
        "isCustom": true,
        "label": "Custom",
        "showTotals": true,
        "type": "custom",
        "uid": "customfield",
      },
      Object {
        "evaluateType": "number",
        "expression": "$customfield + 1",
        "expressionReversed": "$customfield=$_ - 1",
        "isCustom": true,
        "label": "Custom Eval",
        "showTotals": true,
        "type": "customeval",
        "uid": "customfieldeval",
      },
    ]
  }
  methods={
    Object {
      "handleMoveFocusToTargetCell": [MockFunction],
      "handleValueAccept": [MockFunction],
      "handleValueChange": [MockFunction],
      "handleValueReset": [MockFunction],
      "setNavigationMode": [MockFunction],
    }
  }
  navigationMode={true}
  rowIndex={1}
  style={Object {}}
  tableData={
    Array [
      Object {
        "count": 3,
        "hiddenDemo": "Hide me",
        "hiddenDemoNumber": 3,
        "multiplier": 2,
        "name": "Foo",
        "natid": "100a",
        "readOnly": "Can't touch me",
        "select": "a",
        "toggle": true,
        "uid": "a",
      },
      Object {
        "count": "99",
        "def": 3,
        "hiddenDemo": "Hide me",
        "hiddenDemoNumber": 3,
        "name": "Foobar",
        "natid": "10a",
        "uid": "ab",
      },
      Object {
        "description": "A really really really long description that needs to fit into a little box!",
        "hiddenDemo": "Hide me",
        "hiddenDemoNumber": 3,
        "name": "Bar",
        "natid": "50a",
        "uid": "b",
      },
      Object {
        "count": 3,
        "expression": 9,
        "hiddenDemo": "Hide me",
        "hiddenDemoNumber": 3,
        "multiplier": 3,
        "name": "C",
        "uid": "c",
      },
      Object {
        "hiddenDemo": "Hide me",
        "hiddenDemoNumber": 3,
        "name": "",
        "uid": "d",
      },
    ]
  }
>
  <DataTableCellHoverWrap
    className=""
    columnWidth={100}
    disabled={false}
    handleHover={[Function]}
    style={Object {}}
  >
    <div
      className="dataTableCell_wrap  "
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "width": 100,
        }
      }
    >
      <CellNavigationCellWrap
        cellWrapNavBtnRef={
          Object {
            "current": <div
              class=" notFocused row_index_1 column_index_1 column_id_name navigationButton cellWrapBtn button-reset"
              role="presentation"
              style="width: 100%;"
              tabindex="11"
              type="button"
            >
              <button
                class="button-reset hidden"
                style="width: 8px; font-size: 1rem; line-height: 1rem; color: white; border-radius: 0.3rem; font-weight: 800; text-align: center; cursor: pointer;"
                type="button"
              >
                 
              </button>
              <div
                style="height: 100%;"
              >
                <div
                  class="popupMenu_righClickWrap false"
                  style="width: 100%; height: 100%;"
                >
                  <div
                    class="dataTableCellData dataTableCellData-text"
                  >
                    <input
                      class="cellInput-text"
                      style="display: none;"
                      type="text"
                      value="Foobar"
                    />
                    <textarea
                      class="cellInput-textarea"
                      cols="20"
                      rows="10"
                      style="display: none; position: absolute; width: 100%; min-height: 5rem; min-width: 20rem; left: 0px; top: 0px; white-space: normal; resize: none; z-index: 10; overflow: hidden;"
                      type="text"
                      wrap="hard"
                    >
                      Foobar
                    </textarea>
                    <div
                      class="dataTableCellData_text dataTableCellData-text"
                    >
                      Foobar
                    </div>
                  </div>
                </div>
                <div>
                  Missing Popup Root
                </div>
              </div>
            </div>,
          }
        }
        classNames=" notFocused row_index_1 column_index_1 column_id_name"
        columnIndex={1}
        onClick={[Function]}
        onKeyDown={[Function]}
        rowIndex={1}
      >
        <div
          className=" notFocused row_index_1 column_index_1 column_id_name navigationButton cellWrapBtn button-reset"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="presentation"
          style={
            Object {
              "width": "100%",
            }
          }
          tabIndex="11"
          type="button"
        >
          <CellInfoBtn
            message=""
            styles={Object {}}
          >
            <button
              className="button-reset hidden"
              onClick={[Function]}
              style={
                Object {
                  "borderRadius": "0.3rem",
                  "color": "white",
                  "cursor": "pointer",
                  "fontSize": "1rem",
                  "fontWeight": 800,
                  "lineHeight": "1rem",
                  "textAlign": "center",
                  "width": "8px",
                }
              }
              type="button"
            >
               
            </button>
          </CellInfoBtn>
          <DataTableDataCell
            acceptValue={[Function]}
            cellData="Foobar"
            classNames=" notFocused row_index_1 column_index_1 column_id_name"
            columnData={
              Object {
                "label": "Name",
                "required": true,
                "type": "text",
                "uid": "name",
                "unique": true,
              }
            }
            columnId="name"
            componentMap={
              Object {
                "bool": [Function],
                "button": [Function],
                "number": [Function],
                "select": [Function],
                "text": [Function],
                "textLong": [Function],
                "toggle": [Function],
                "undefined": [Function],
              }
            }
            editMode={false}
            focused={false}
            isDisabled={false}
            resetValue={[Function]}
            rowData={
              Object {
                "count": "99",
                "def": 3,
                "hiddenDemo": "Hide me",
                "hiddenDemoNumber": 3,
                "name": "Foobar",
                "natid": "10a",
                "uid": "ab",
              }
            }
            rowId="ab"
            rowIndex={1}
            updateData={[Function]}
          >
            <div
              style={
                Object {
                  "height": "100%",
                }
              }
            >
              <CellRightClickWrapper
                clearCell={[Function]}
                readOnly={false}
                setAsDefault={[Function]}
              >
                <RightClickWrapper
                  items={
                    Array [
                      Object {
                        "label": "Clear",
                        "onClick": [Function],
                        "uid": "clearCell",
                      },
                      Object {
                        "label": "Set as Default",
                        "onClick": [Function],
                        "uid": "setDefault",
                      },
                    ]
                  }
                  popupRoot="root"
                >
                  <div
                    className="popupMenu_righClickWrap false"
                    onContextMenu={[Function]}
                    style={
                      Object {
                        "height": "100%",
                        "width": "100%",
                      }
                    }
                  >
                    <Component
                      acceptValue={[Function]}
                      cellData="Foobar"
                      classNames=" notFocused row_index_1 column_index_1 column_id_name"
                      columnData={
                        Object {
                          "label": "Name",
                          "required": true,
                          "type": "text",
                          "uid": "name",
                          "unique": true,
                        }
                      }
                      columnId="name"
                      componentMap={
                        Object {
                          "bool": [Function],
                          "button": [Function],
                          "number": [Function],
                          "select": [Function],
                          "text": [Function],
                          "textLong": [Function],
                          "toggle": [Function],
                          "undefined": [Function],
                        }
                      }
                      editMode={false}
                      focused={false}
                      isDisabled={false}
                      resetValue={[Function]}
                      rowData={
                        Object {
                          "count": "99",
                          "def": 3,
                          "hiddenDemo": "Hide me",
                          "hiddenDemoNumber": 3,
                          "name": "Foobar",
                          "natid": "10a",
                          "uid": "ab",
                        }
                      }
                      rowId="ab"
                      rowIndex={1}
                      updateData={[Function]}
                    >
                      <DataTableCellText
                        acceptValue={[Function]}
                        cellData="Foobar"
                        classNames=" notFocused row_index_1 column_index_1 column_id_name"
                        columnData={
                          Object {
                            "label": "Name",
                            "required": true,
                            "type": "text",
                            "uid": "name",
                            "unique": true,
                          }
                        }
                        columnId="name"
                        componentMap={
                          Object {
                            "bool": [Function],
                            "button": [Function],
                            "number": [Function],
                            "select": [Function],
                            "text": [Function],
                            "textLong": [Function],
                            "toggle": [Function],
                            "undefined": [Function],
                          }
                        }
                        editMode={false}
                        focused={false}
                        isDisabled={false}
                        resetValue={[Function]}
                        rowData={
                          Object {
                            "count": "99",
                            "def": 3,
                            "hiddenDemo": "Hide me",
                            "hiddenDemoNumber": 3,
                            "name": "Foobar",
                            "natid": "10a",
                            "uid": "ab",
                          }
                        }
                        rowId="ab"
                        rowIndex={1}
                        updateData={[Function]}
                      >
                        <div
                          className="dataTableCellData dataTableCellData-text"
                        >
                          <input
                            className="cellInput-text"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onKeyDown={[Function]}
                            style={
                              Object {
                                "display": "none",
                              }
                            }
                            type="text"
                            value="Foobar"
                          />
                          <textarea
                            className="cellInput-textarea"
                            cols="20"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onKeyDown={[Function]}
                            rows="10"
                            style={
                              Object {
                                "display": "none",
                                "left": 0,
                                "minHeight": "5rem",
                                "minWidth": "20rem",
                                "overflow": "hidden",
                                "position": "absolute",
                                "resize": "none",
                                "top": 0,
                                "whiteSpace": "normal",
                                "width": "100%",
                                "zIndex": 10,
                              }
                            }
                            type="text"
                            value="Foobar"
                            wrap="hard"
                          />
                          <div
                            className="dataTableCellData_text dataTableCellData-text"
                          >
                            Foobar
                          </div>
                        </div>
                      </DataTableCellText>
                    </Component>
                  </div>
                  <PopupMenu
                    isOpenOverride={false}
                    items={
                      Array [
                        Object {
                          "label": "Clear",
                          "onClick": [Function],
                          "uid": "clearCell",
                        },
                        Object {
                          "label": "Set as Default",
                          "onClick": [Function],
                          "uid": "setDefault",
                        },
                      ]
                    }
                    onCloseCallback={[Function]}
                    popupRoot="root"
                    position={
                      Object {
                        "x": 0,
                        "y": 0,
                      }
                    }
                  >
                    <div>
                      Missing Popup Root
                    </div>
                  </PopupMenu>
                </RightClickWrapper>
              </CellRightClickWrapper>
            </div>
          </DataTableDataCell>
        </div>
      </CellNavigationCellWrap>
    </div>
  </DataTableCellHoverWrap>
</Cell>
`;
